# Spark DataFrames

In [1]:
df_from_other_list = spark.createDataFrame([('Chris',67),('Logan',70)], ['name','score'])
df_from_other_list.show()

VBox()

Starting Spark application


SparkSession available as 'spark'.
----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 53230)
----------------------------------------
+-----+-----+
| name|score|
+-----+-----+
|Chris|   67|
|Logan|   70|
+-----+-----+

Traceback (most recent call last):
  File "/usr/lib64/python2.7/SocketServer.py", line 293, in _handle_request_noblock
    self.process_request(request, client_address)
  File "/usr/lib64/python2.7/SocketServer.py", line 321, in process_request
    self.finish_request(request, client_address)
  File "/usr/lib64/python2.7/SocketServer.py", line 334, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib64/python2.7/SocketServer.py", line 655, in __init__
    self.handle()
  File "/usr/lib/spark/python/lib/pyspark.zip/pyspark/accumulators.py", line 266, in handle
    poll(authenticate_and_accum_updates)
  File "/usr/lib/spark/python/lib/pyspark.zip/pyspark/accumulators.py", lin

In [2]:
from pyspark.sql.types import FloatType
df_from_list = spark.createDataFrame([1.0,2.0,3.0,4.0,5.0], FloatType())
df_from_list.show()

VBox()

+-----+
|value|
+-----+
|  1.0|
|  2.0|
|  3.0|
|  4.0|
|  5.0|
+-----+

In [3]:
lot_rdd = sc.parallelize([('Chris',67),('Logan',70)])

VBox()

In [4]:
dfPeople = spark.createDataFrame(lot_rdd)
dfPeople.show()

VBox()

+-----+---+
|   _1| _2|
+-----+---+
|Chris| 67|
|Logan| 70|
+-----+---+

In [5]:
from pyspark.sql import Row
lot_rdd_named_columns = lot_rdd.map(lambda x: Row(name=x[0], score=int(x[1])))
dfPeople_named_columns = spark.createDataFrame(lot_rdd_named_columns)
dfPeople_named_columns.show()

VBox()

+-----+-----+
| name|score|
+-----+-----+
|Chris|   67|
|Logan|   70|
+-----+-----+

In [6]:
df = spark.read.json('s3://umsi-data-science/data/yelp/business.json')

VBox()

In [7]:
df.take(5)

VBox()

[Row(address=u'4855 E Warner Rd, Ste B9', attributes=Row(AcceptsInsurance=True, AgesAllowed=None, Alcohol=None, Ambience=None, BYOB=None, BYOBCorkage=None, BestNights=None, BikeParking=None, BusinessAcceptsBitcoin=None, BusinessAcceptsCreditCards=True, BusinessParking=None, ByAppointmentOnly=True, Caters=None, CoatCheck=None, Corkage=None, DietaryRestrictions=None, DogsAllowed=None, DriveThru=None, GoodForDancing=None, GoodForKids=None, GoodForMeal=None, HairSpecializesIn=None, HappyHour=None, HasTV=None, Music=None, NoiseLevel=None, Open24Hours=None, OutdoorSeating=None, RestaurantsAttire=None, RestaurantsCounterService=None, RestaurantsDelivery=None, RestaurantsGoodForGroups=None, RestaurantsPriceRange2=None, RestaurantsReservations=None, RestaurantsTableService=None, RestaurantsTakeOut=None, Smoking=None, WheelchairAccessible=None, WiFi=None), business_id=u'FYWN1wneV18bWNgQjJ2GNg', categories=[u'Dentists', u'General Dentistry', u'Health & Medical', u'Oral Surgeons', u'Cosmetic Denti

In [8]:
df.printSchema()

VBox()

root
 |-- address: string (nullable = true)
 |-- attributes: struct (nullable = true)
 |    |-- AcceptsInsurance: boolean (nullable = true)
 |    |-- AgesAllowed: string (nullable = true)
 |    |-- Alcohol: string (nullable = true)
 |    |-- Ambience: struct (nullable = true)
 |    |    |-- casual: boolean (nullable = true)
 |    |    |-- classy: boolean (nullable = true)
 |    |    |-- divey: boolean (nullable = true)
 |    |    |-- hipster: boolean (nullable = true)
 |    |    |-- intimate: boolean (nullable = true)
 |    |    |-- romantic: boolean (nullable = true)
 |    |    |-- touristy: boolean (nullable = true)
 |    |    |-- trendy: boolean (nullable = true)
 |    |    |-- upscale: boolean (nullable = true)
 |    |-- BYOB: boolean (nullable = true)
 |    |-- BYOBCorkage: string (nullable = true)
 |    |-- BestNights: struct (nullable = true)
 |    |    |-- friday: boolean (nullable = true)
 |    |    |-- monday: boolean (nullable = true)
 |    |    |-- saturday: boolean (nullab

In [9]:
df.select("name").show()

VBox()

+--------------------+
|                name|
+--------------------+
|    Dental by Design|
| Stephen Szabo Salon|
|Western Motor Veh...|
|    Sports Authority|
|Brick House Taver...|
|             Messina|
|          BDJ Realty|
|         Soccer Zone|
|    Any Given Sundae|
|Detailing Gone Mo...|
|   East Coast Coffee|
|CubeSmart Self St...|
|T & T Bakery and ...|
|Complete Dental Care|
|Showmars Governme...|
|      Alize Catering|
|      T & Y Nail Spa|
|Meineke Car Care ...|
|Senior's Barber Shop|
|Maxim Bakery & Re...|
+--------------------+
only showing top 20 rows

In [10]:
df.select(df['name'], df['review_count'] + 1).show()

VBox()

+--------------------+------------------+
|                name|(review_count + 1)|
+--------------------+------------------+
|    Dental by Design|                23|
| Stephen Szabo Salon|                12|
|Western Motor Veh...|                19|
|    Sports Authority|                10|
|Brick House Taver...|               117|
|             Messina|                 6|
|          BDJ Realty|                 6|
|         Soccer Zone|                10|
|    Any Given Sundae|                16|
|Detailing Gone Mo...|                 8|
|   East Coast Coffee|                 4|
|CubeSmart Self St...|                24|
|T & T Bakery and ...|                39|
|Complete Dental Care|                 6|
|Showmars Governme...|                 8|
|      Alize Catering|                13|
|      T & Y Nail Spa|                21|
|Meineke Car Care ...|                10|
|Senior's Barber Shop|                66|
|Maxim Bakery & Re...|                35|
+--------------------+------------

In [12]:
df.filter(df['stars'] >= 4).show()

VBox()

+--------------------+--------------------+--------------------+--------------------+------------+--------------------+-------+-------------+--------------+--------------------+---------------+-----------+------------+-----+-----+
|             address|          attributes|         business_id|          categories|        city|               hours|is_open|     latitude|     longitude|                name|   neighborhood|postal_code|review_count|stars|state|
+--------------------+--------------------+--------------------+--------------------+------------+--------------------+-------+-------------+--------------+--------------------+---------------+-----------+------------+-----+-----+
|4855 E Warner Rd,...|[true,,,,,,,,, tr...|FYWN1wneV18bWNgQj...|[Dentists, Genera...|   Ahwatukee|[7:30-17:00, 7:30...|      1|   33.3306902|  -111.9785992|    Dental by Design|               |      85044|          22|  4.0|   AZ|
|      Richterstr. 11|[,, beer_and_wine...|o9eMRCWt5PkpLDE0g...|[Italian, Re

In [13]:
df.groupBy('stars').count().show()

VBox()

+-----+-----+
|stars|count|
+-----+-----+
|  3.5|32038|
|  4.5|24796|
|  2.5|16148|
|  1.0| 3788|
|  4.0|33492|
|  3.0|23142|
|  2.0| 9320|
|  1.5| 4303|
|  5.0|27540|
+-----+-----+

In [14]:
df.groupBy('stars').count().sort('stars', ascending=False).show()

VBox()

+-----+-----+
|stars|count|
+-----+-----+
|  5.0|27540|
|  4.5|24796|
|  4.0|33492|
|  3.5|32038|
|  3.0|23142|
|  2.5|16148|
|  2.0| 9320|
|  1.5| 4303|
|  1.0| 3788|
+-----+-----+

In [15]:
df_from_other_list = spark.createDataFrame([('Chris',[67,42]),('Logan',[70,72])],['name','scores'])

VBox()

In [16]:
df_from_other_list.show()

VBox()

+-----+--------+
| name|  scores|
+-----+--------+
|Chris|[67, 42]|
|Logan|[70, 72]|
+-----+--------+

In [17]:
from pyspark.sql.functions import explode

VBox()

In [18]:
df_exploded = df_from_other_list.withColumn('score',explode('scores'))

VBox()

In [19]:
df_exploded.show()

VBox()

+-----+--------+-----+
| name|  scores|score|
+-----+--------+-----+
|Chris|[67, 42]|   67|
|Chris|[67, 42]|   42|
|Logan|[70, 72]|   70|
|Logan|[70, 72]|   72|
+-----+--------+-----+

In [20]:
import pyspark.sql.functions as F
from pyspark.sql.functions import col
df_exploded.withColumn('good',F.when(df_exploded['score'] > 50,1).otherwise(0)).show()

VBox()

+-----+--------+-----+----+
| name|  scores|score|good|
+-----+--------+-----+----+
|Chris|[67, 42]|   67|   1|
|Chris|[67, 42]|   42|   0|
|Logan|[70, 72]|   70|   1|
|Logan|[70, 72]|   72|   1|
+-----+--------+-----+----+

In [21]:
STOPWORDS = {'i', 'we', 'ourselves', 'hers', 'between', 'yourself', 'but', 'again', 'there', 'about', 'once', 'during', 'out', 'very', 'having', 'with', 'they', 'own', 'an', 'be', 'some', 'for', 'do', 'its', 'yours', 'such', 'into', 'of', 'most', 'itself', 'other', 'off', 'is', 's', 'am', 'or', 'who', 'as', 'from', 'him', 'each', 'the', 'themselves', 'until', 'below', 'are', 'we', 'these', 'your', 'his', 'through', 'don', 'nor', 'me', 'were', 'her', 'more', 'himself', 'this', 'down', 'should', 'our', 'their', 'while', 'above', 'both', 'up', 'to', 'ours', 'had', 'she', 'all', 'no', 'when', 'at', 'any', 'before', 'them', 'same', 'and', 'been', 'have', 'in', 'will', 'on', 'does', 'yourselves', 'then', 'that', 'because', 'what', 'over', 'why', 'so', 'can', 'did', 'not', 'now', 'under', 'he', 'you', 'herself', 'has', 'just', 'where', 'too', 'only', 'myself', 'which', 'those', 'i', 'after', 'few', 'whom', 't', 'being', 'if', 'theirs', 'my', 'against', 'a', 'by', 'doing', 'it', 'how', 'further', 'was', 'here', 'than'}
sw = sc.broadcast(STOPWORDS)

VBox()

## Text analysis of Yelp reviews, warm-up

In [22]:
# This takes about 3 minutes (180s) on a single-node m4.xlarge cluster
df = spark.read.json('s3://umsi-data-science/data/yelp/review.json.gz')

VBox()

In [23]:
df.printSchema()

VBox()

root
 |-- business_id: string (nullable = true)
 |-- cool: long (nullable = true)
 |-- date: string (nullable = true)
 |-- funny: long (nullable = true)
 |-- review_id: string (nullable = true)
 |-- stars: long (nullable = true)
 |-- text: string (nullable = true)
 |-- useful: long (nullable = true)
 |-- user_id: string (nullable = true)

### Q1: How many reviews are there?

In [25]:
df.count()

VBox()

5261669

There are 5261669 reviews

### Q2: How many positive (i.e. 4- or 5-star) reviews are there? 

In [26]:
df.filter(df['stars'] >= 4).count()

VBox()

3476664

There are 3476664 4 or 5 star reviews